### ISSUE TITLE
Add USB_F_MIDI Kernel Module

### ISSUE THREAD

**Issue by simotek** [jamesMcMeex/m8c-rg35xx-knulli#7]
Hi, I've been playing with your work on the rg35xx-plus, but mostly for LittleGPTracker.

One question I have is would it be possible to build the USB MIDI kernel module using your script? I had a shot but my knowledge of the kernel build system isn't good enough to figure it out.

**Comment by DRCRecoveryData** [comment]
> Hi, I've been playing with your work on the rg35xx-plus, but mostly for LittleGPTracker.
> 
> One question I have is would it be possible to build the USB MIDI kernel module using your script? I had a shot but my knowledge of the kernel build system isn't good enough to figure it out.

Yes, it is possible to build the USB MIDI kernel module, but it requires some adjustments to the kernel configuration and build process. You would need to enable the necessary kernel modules and rebuild the kernel to include USB MIDI support.

Here's a general approach you can take to achieve this:

### Steps to Enable USB MIDI in the Kernel:

1. **Modify the Kernel Configuration:**
   You'll need to make sure the required kernel modules for USB MIDI are enabled in the kernel config.

   In your case, since you're working with the `linux-sunxi64-legacy.config`, you need to enable the USB MIDI module in that configuration file or during the kernel build process.

   Add or modify these lines in your kernel configuration (`linux-sunxi64-legacy.config`):
   ```bash
   CONFIG_SND=y
   CONFIG_SND_RAWMIDI=y
   CONFIG_SND_SEQ_DEVICE=y
   CONFIG_SND_USB=y
   CONFIG_SND_USB_AUDIO=m
   CONFIG_SND_USB_MIDI=m
   ```

   If these options are set to `m`, the modules will be built as loadable kernel modules, which is likely what you want for USB MIDI.

2. **Prepare for Kernel Build:**
   In your `build_script.sh`, after downloading the kernel and preparing the environment, you can include steps to ensure that the configuration is applied:

   ```bash
   # Apply the custom config for enabling USB MIDI
   cp /build/linux-sunxi64-legacy.config /build/linux-4.9.170/.config
   cd /build/linux-4.9.170

   # Update configuration to enable USB MIDI
   make ARCH=arm64 oldconfig
   ```

3. **Build the Kernel and Modules:**
   Once the configuration is updated, you can build the kernel and the specific USB MIDI modules:

   ```bash
   # Build the kernel and the modules
   make ARCH=arm64 CROSS_COMPILE=aarch64-linux-gnu- -j$(nproc) zImage modules

   # Install the modules
   make ARCH=arm64 CROSS_COMPILE=aarch64-linux-gnu- modules_install INSTALL_MOD_PATH=/path/to/your/output
   ```

4. **Load the USB MIDI Module:**
   After the build is complete and the kernel is flashed onto your device, you can load the `snd-usb-audio` module to activate USB MIDI support:

   ```bash
   modprobe snd-usb-audio
   ```

   Or, if you've built the module as part of the kernel image, it should be available automatically once the device detects a USB MIDI device.

### Update Your Build Script:
You may need to extend the `build_script.sh` to handle the kernel configuration and the module installation process. You could also integrate the steps above into your Dockerfile if you want to automate the entire process.

If you'd like more specific guidance or if you have a more detailed build script, feel free to share it and I can help tailor the instructions!

**Comment by jamesMcMeex** [comment]
@simotek Any progress on this? I too, am not knowledgable on the build process TBH!

**Comment by jamesMcMeex** [comment]
Closing, hope all is good :)
